make: gui_ prerendering

oharboe opened the pull request:
less [click and wait](https://github.com/The-OpenROAD-Project/OpenROAD/issues/6074) when the GUI opens

Clock tree prerendering is still missing, because there is  no way yet to disable the clock tree rendering from .tcl after   the clock tree has been rendered.
    
Some progress added so that the user can tell what it is  that is taking time. Also added a final gui::unminimize  log item so the user knows when the GUI should have unminimized(which it doesn't on Ubuntu 24.04 with Wayland yet, known issue filed).

```    
make gui_route
[deleted]
estimate_parasitics -global_routing
Populating timing paths...OK
Prerendering Placement heatmap...
Prerendering Routing heatmap...
Prerendering RUDY heatmap...
Prerendering Power heatmap...
gui::select_chart "Endpoint Slack"
gui::update_timing_report
gui::unminimize
```

oharboe committed:
make: gui_ prerendering

less click and wait when the GUI opens for large designs

clock tree prerendering is still missing, because there is
no way yet to disable the clock tree rendering from .tcl after
the clock tree has been rendered.

Some progress added so that the user can tell what it is
that is taking time. Also added a final gui::unminimize
log item so the user knows when the GUI should have
unminimized(which it doesn't on Ubuntu 24.04 with Wayland
yet, known issue filed).

make gui_route
[deleted]
estimate_parasitics -global_routing
Populating timing paths...OK
Prerendering Placement heatmap...
Prerendering Routing heatmap...
Prerendering RUDY heatmap...
Prerendering Power heatmap...
gui::select_chart "Endpoint Slack"
gui::update_timing_report
gui::unminimize

https://github.com/The-OpenROAD-Project/OpenROAD/issues/6074

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe commented on flow/scripts/open.tcl:
```diff
@@ -60,12 +60,43 @@ proc read_timing {input_file} {
+      gui::select_clockviewer_clock $clock_name
+      save_image -resolution $resolution $::env(REPORTS_DIR)/cts_${clock_name}_layout.webp
+    }
+  }
```
@maliberty How do I unselect the clock tree and return the GUI to the default view?

`make gui_cts`

![image](https://github.com/user-attachments/assets/206cd8f8-ad81-4235-8bb2-c7f62a0675b5)

oharboe commented on flow/scripts/open.tcl:
```diff
@@ -80,12 +80,19 @@ if {[env_var_equals GUI_TIMING 1]} {
     }
+    if {[string equal $heatmap IRDrop] && ![env_var_exists_and_non_empty PWR_NETS_VOLTAGES]} {
+      puts "Skipping $heatmap heatmap, no PWR_NETS_VOLTAGES available"
+      continue
```
@maliberty How do I skip the IRDrop heatmap when there is no data?

```
 make gui_floorplan
ODB_FILE=./results/nangate45/gcd/base/2_floorplan.odb GUI_SHOW=1 /home/oyvind/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad -gui  -minimize /home/oyvind/OpenROAD-flow-scripts/flow/scripts/open.tcl
QSocketNotifier: Can only be used with threads started with QThread
OpenROAD v2.0-17013-gf7f634f88 
Features included (+) or not (-): +Charts +GPU +GUI +Python
This program is licensed under the BSD-3 license. See the LICENSE file for details.
Components of this program may be licensed under more restrictive licenses which must be honored.
GUI_TIMING=1 reading timing, takes a little while for large designs...
Populating timing paths...OK
Prerendering...
Dumping Placement heatmap...
Dumping IRDrop heatmap...
[ERROR GUI-0072] "IR Drop" is not populated with data.
[ERROR GRT-0010] Instance ctrl.state.out\[0\]$_DFF_P_ is not placed.
[ERROR GUI-0070] Error: open.tcl, 101 GRT-0010
^Cmake: *** [Makefile:1013: gui_2_floorplan.odb] Interrupt
```

oharboe replied:
How do I skip Placement heatmap?

[ERROR GRT-0010] Instance ctrl.state.out\[0\]$_DFF_P_ is not placed.
[ERROR GUI-0070] Error: open.tcl, 105 GRT-0010

maliberty replied:
We don't save the ir drop information in odb so if you are reloading there isn't any.

oharboe replied:
I don't know what that means... I haven't used IR Drop yet. I'm going to remove IRDrop for now if there isn't an easy way to skip it when prerendering should be skipped.

Placement heatmap is more interesting: how do I skip it when there's no placement?

maliberty replied:
You could loop over the instances and see if any are placed.

oharboe replied:
Worked. I haven't tested how well it scales though.

oharboe committed:
openroad: bump

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

maliberty commented on flow/scripts/open.tcl:
```diff
@@ -60,12 +60,74 @@ proc read_timing {input_file} {
+  #         $::env(OBJECTS_DIR)/dummy.png
+  #     break
+  #   }
+  # }
```
open should not write files as a side effect.  This should be removed

oharboe replied:
Agreed, but how else to precalculate this?

maliberty replied:
Does it actually save time?  Most of the effort is to get sta warmed up.  Drawing the tree should be cheap.

oharboe replied:
Yes, last I checked.

However, I can delete the commented code and if this PR is otherwise good to go, we can square away what we already have a solution for and I can then create a new PR with a narrower concern and some numbers.

A way to precalculate this that does not modify GUI state and also does not write files to disk is needed, which isnt available yet, right?

maliberty replied:
Yes, if there really is a need.  I would be curious to see the time saved.

oharboe committed:
make: gui_ prerendering

delete clock tree commented out code for now,
revisit later.

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
make: gui_ prerendering simplification

simplify now that grt::have_routes checks if placement
is done first.

no need to save dummy.png to prerender.

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

maliberty commented on flow/scripts/open.tcl:
```diff
@@ -60,12 +60,62 @@ proc read_timing {input_file} {
+    }
+  }
+
+  set have_routes [expr {$placed && [grt::have_routes]}]
```
Do we still need to compute/check $placed?  I thought this was fixed.

oharboe replied:
Yes: there is no ppl::have_placed method. I can simplify the code a bit here though for the have_routes case.

maliberty commented on flow/scripts/open.tcl:
```diff
@@ -60,12 +60,62 @@ proc read_timing {input_file} {
+    }
+    puts "Prerendering $heatmap heatmap..."
+    gui::set_heatmap $heatmap rebuild 1
+    gui::dump_heatmap $heatmap $::env(REPORTS_DIR)/dummy.png
```
Can we avoid the file side effect?

oharboe replied:
yes: fixed

oharboe committed:
Merge remote-tracking branch 'origin/master' into HEAD

oharboe committed:
Merge remote-tracking branch 'origin/master' into HEAD

oharboe committed:
makefile: make gui runs gui::update_timing_report if GUI_TIMING=1

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
Merge remote-tracking branch 'origin/master' into HEAD
